### PR TITLE
Fix allocator bug

### PR DESCRIPTION
**Pull request by sneaxiy**
The previous architecture of `Allocator` would cause some `Allocation` cannot be released, saying potential memory leak. This PR fixes it.

For more details of Allocator design, please refer to `paddle/fluid/memory/allocation/allocator.h` in this PR.

CE has passed. Please see [here](http://ce.paddlepaddle.org:8080/viewLog.html?buildId=43441&tab=buildResultsDiv&buildTypeId=PaddleModesl_ModelDebug_Model11Build).

### CODE REVIEW

File: paddle/fluid/memory/allocation/allocator.h [hunk 1]
```diff
@@ -26,40 +28,73 @@ namespace allocation {
...
  public:
-  Allocation(void* ptr, size_t size, platform::Place place)
-      : allocator_(nullptr), ptr_(ptr), size_(size), place_(place) {}
+  inline Allocation(void* ptr, size_t size, platform::Place place)
+      : ptr_(ptr), size_(size), place_(place) {}
 
   Allocation(const Allocation& o) = delete;
   Allocation& operator=(const Allocation& o) = delete;
+  Allocation(Allocation&& o) = delete;
+  Allocation& operator=(Allocation&& o) = delete;
```
Comment (chengduoZH) [inline paddle/fluid/memory/allocation/allocator.h:90]: Why remove those methods?

Comment (sneaxiy) [reply paddle/fluid/memory/allocation/allocator.h:90]: Because `Allocation` should not be moved.

**Review by chengduoZH — COMMENTED** [review]
Please add doc to explain the reason for adding `InlinedVector` in this PR.

**Review by chengduoZH — APPROVED** [review]
LGTM

File: paddle/fluid/memory/allocation/allocator.h (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/memory/allocation/allocator.h:90]: Because `Allocation` should not be moved. So I remove those methods generated by compiler.

**Comment by sneaxiy** [comment]
@chengduoZH I would add some doc of `InlinedVector` in the next PR asap.

**Pull request merged**